src/main/java/com/epam/service/impl/TrainerServiceImpl.java

```diff
@@ -2,68 +2,80 @@
+    public boolean updateTrainer(Trainer trainer) {
+        checkUsernameAndPasswordMatching(trainer.getUser().getUsername(), trainer.getUser().getPassword());
+        return trainerDao.update(trainer.getId(), trainer);
     }
 
     public void deleteTrainer(Integer trainerId) {
-        trainerDao.delete(TrainerDao.TRAINERS, trainerId);
+        Trainer trainer = getTrainerById(trainerId);
+        checkUsernameAndPasswordMatching(trainer.getUser().getUsername(), trainer.getUser().getPassword());
+        trainerDao.delete(trainerId);
     }
 
     public Trainer getTrainerById(Integer trainerId) {
-        return trainerDao.findById(TrainerDao.TRAINERS, trainerId);
+        Trainer trainer = trainerDao.findById(trainerId);
+        checkUsernameAndPasswordMatching(trainer.getUser().getUsername(), trainer.getUser().getPassword());
+        return trainer;
     }
 
     public List<Trainer> getAllTrainer() {
-        return trainerDao.findAll(TrainerDao.TRAINERS).values()
+        return trainerDao.findAll()
                 .stream()
                 .toList();
     }
+    public boolean changePassword(Integer trainerId,String currentPassword, String newPassword) {
+        Trainer trainer = trainerDao.findById(trainerId);
+        checkUsernameAndPasswordMatching(trainer.getUser().getUsername(), trainer.getUser().getPassword());
+
+        if (currentPassword.equals(trainer.getUser().getPassword())) {
+            trainer.getUser().setPassword(newPassword);
+            trainerDao.save(trainer);
+            return true;
+        }
+        return false;
+    }
 
-    public String generateTrainerUsername(Trainer trainee) {
-        String baseUsername = trainee.getUser().getFirstName() + "." + trainee.getUser().getLastName();
-        String username = baseUsername;
+    public void activateDeactivateTrainer(Integer trainerId, boolean activate) {
+        Trainer trainer = trainerDao.findById(trainerId);
+        User user = trainer.getUser();
+        checkUsernameAndPasswordMatching(user.getUsername(), user.getPassword());
+        user.setActive(activate);
+        trainerDao.save(trainer);
 
-        Set<String> existingUsernames = new HashSet<>();
-        trainerDao.findAll(TrainerDao.TRAINERS).values().forEach(existingTrainee -> {
-            String existingName = existingTrainee.getUser().getFirstName() + "." + existingTrainee.getUser().getLastName();
-            existingUsernames.add(existingName);
-        });
+    }
 
-        // Add serial number if the username already exists
-        int serialNumber = 1;
-        while (existingUsernames.contains(username)) {
-            username = baseUsername + serialNumber;
-            serialNumber++;
+    public void checkUsernameAndPasswordMatching(String username, String password) {
+        Trainer trainer = trainerDao.findByUsername(username);
+        if (trainer == null || !trainer.getUser().getPassword().equals(password)) {
```

Comment:
What if `trainer.getUser()` returns null? What if `trainer.getUser().getPassword()` returns null? Is it possible in your implementation? Even if you think it's impossible please take care about it in this method. And add appropriate test cases.